In [1]:
#imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

import glob
import os

In [2]:
#change the settings for good output
pd.set_option('display.max_rows',100)
pd.set_option('display.max_columns',20)
pd.set_option('display.max_colwidth',200)

In [3]:
#read the file names and join with its extension (.json)
path = 'C:\Data-Analysis-projects\FC Barcelona Data Analysis\Datasets'
file_name = glob.glob(os.path.join(path,'*.json'))
print(file_name)

['C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20042005-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20052006-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20062007-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20072008-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20082009-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20092010-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20102011-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20112012-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20122013-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data Analysis\\Datasets\\20132014-FCBarcelona.json', 'C:\\Data-Analysis-projects\\FC Barcelona Data An

In [4]:
#read the file in pandas and concatenate them to a single dataframe
dfs = []
for file in file_name:
    df = pd.read_json(file)
    dfs.append(df)
    
fcb = pd.concat(dfs, axis=0, ignore_index=True)

In [8]:
#first 5 data
fcb.head(3)

match_id  match_date      kick_off  \
0     68314  2004-12-04  20:00:00.000   
1     68313  2004-10-24  21:00:00.000   
2     68316  2005-05-01  19:00:00.000   

                                                                      competition  \
0  {'competition_id': 11, 'country_name': 'Spain', 'competition_name': 'La Liga'}   
1  {'competition_id': 11, 'country_name': 'Spain', 'competition_name': 'La Liga'}   
2  {'competition_id': 11, 'country_name': 'Spain', 'competition_name': 'La Liga'}   

                                          season  \
0  {'season_id': 37, 'season_name': '2004/2005'}   
1  {'season_id': 37, 'season_name': '2004/2005'}   
2  {'season_id': 37, 'season_name': '2004/2005'}   

                                                                                                                                                                                                 home_team  \
0  {'home_team_id': 217, 'home_team_name': 'Barcelona', 'home_team_gender': 'male', 'home_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 4953, 'name': 'Frank Rijkaard...   
1  {'home_team_id': 217, 'home_team_name': 'Barcelona', 'home_team_gender': 'male', 'home_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 4953, 'name': 'Frank Rijkaard...   
2  {'home_team_id': 217, 'home_team_name': 'Barcelona', 'home_team_gender': 'male', 'home_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 4953, 'name': 'Frank Rijkaard...   

                                                                                                                                                                                                 away_team  \
0  {'away_team_id': 223, 'away_team_name': 'Málaga', 'away_team_gender': 'male', 'away_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 4994, 'name': 'Gregorio Manzano ...   
1  {'away_team_id': 422, 'away_team_name': 'Osasuna', 'away_team_gender': 'male', 'away_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 3196, 'name': 'Javier Aguirre O...   
2  {'away_team_id': 608, 'away_team_name': 'Albacete', 'away_team_gender': 'male', 'away_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 6233, 'name': 'Martín Monteagu...   

   home_score  away_score match_status match_status_360      last_updated  \
0           4           0    available        scheduled  2020-07-29T05:00   
1           3           0    available        scheduled  2020-07-29T05:00   
2           2           0    available        scheduled  2020-07-29T05:00   

          last_updated_360  \
0  2021-06-13T16:17:31.694   
1  2021-06-13T16:17:31.694   
2  2021-06-13T16:17:31.694   

                                                                              metadata  \
0  {'data_version': '1.1.0', 'shot_fidelity_version': '2', 'xy_fidelity_version': '2'}   
1  {'data_version': '1.1.0', 'shot_fidelity_version': '2', 'xy_fidelity_version': '2'}   
2  {'data_version': '1.1.0', 'shot_fidelity_version': '2', 'xy_fidelity_version': '2'}   

   match_week                    competition_stage  \
0          14  {'id': 1, 'name': 'Regular Season'}   
1           8  {'id': 1, 'name': 'Regular Season'}   
2          34  {'id': 1, 'name': 'Regular Season'}   

                                                                    stadium  \
0  {'id': 342, 'name': 'Camp Nou', 'country': {'id': 214, 'name': 'Spain'}}   
1  {'id': 342, 'name': 'Camp Nou', 'country': {'id': 214, 'name': 'Spain'}}   
2  {'id': 342, 'name': 'Camp Nou', 'country': {'id': 214, 'name': 'Spain'}}   

                                                                                       referee  
0           {'id': 993, 'name': 'José Omar Losantos', 'country': {'id': 214, 'name': 'Spain'}}  
1  {'id': 994, 'name': 'Vicente José Lizondo Cortés', 'country': {'id': 214, 'name': 'Spain'}}  
2     {'id': 2535, 'na

In [6]:
#basic info like the shape, datatypes and null values
fcb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 520 entries, 0 to 519
Data columns (total 18 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   match_id           520 non-null    int64 
 1   match_date         520 non-null    object
 2   kick_off           520 non-null    object
 3   competition        520 non-null    object
 4   season             520 non-null    object
 5   home_team          520 non-null    object
 6   away_team          520 non-null    object
 7   home_score         520 non-null    int64 
 8   away_score         520 non-null    int64 
 9   match_status       520 non-null    object
 10  match_status_360   520 non-null    object
 11  last_updated       520 non-null    object
 12  last_updated_360   520 non-null    object
 13  metadata           520 non-null    object
 14  match_week         520 non-null    int64 
 15  competition_stage  520 non-null    object
 16  stadium            520 non-null    object
 1

> The dataset has 520 matches played with 18 columns information, only referee has missing values and some datatypes need to change

In [7]:
# descriptive statistics
fcb.describe()

match_id  home_score  away_score  match_week
count  5.200000e+02  520.000000  520.000000  520.000000
mean   3.744410e+05    1.982692    1.500000   19.457692
std    9.194129e+05    1.723041    1.492434   10.777579
min    9.575000e+03    0.000000    0.000000    1.000000
25%    6.915675e+04    1.000000    0.000000   10.000000
50%    6.932250e+04    2.000000    1.000000   20.000000
75%    2.667258e+05    3.000000    2.000000   29.000000
max    3.773695e+06    8.000000    8.000000   38.000000

> We can see the maximum score is 8 goals, minimum is obviously zero, and an average of 2 goals precisely. 

In [27]:
#check for duplicates
fcb['match_id'].duplicated().any()

False

In [9]:
#check the competition column
fcb['competition'].value_counts()

{'competition_id': 11, 'country_name': 'Spain', 'competition_name': 'La Liga'}    520
Name: competition, dtype: int64

> it's has only a single value object. it should be discarded

In [11]:
#check the season column
fcb['season'].value_counts()

{'season_id': 26, 'season_name': '2014/2015'}    38
{'season_id': 23, 'season_name': '2011/2012'}    37
{'season_id': 1, 'season_name': '2017/2018'}     36
{'season_id': 90, 'season_name': '2020/2021'}    35
{'season_id': 21, 'season_name': '2009/2010'}    35
{'season_id': 4, 'season_name': '2018/2019'}     34
{'season_id': 2, 'season_name': '2016/2017'}     34
{'season_id': 42, 'season_name': '2019/2020'}    33
{'season_id': 22, 'season_name': '2010/2011'}    33
{'season_id': 27, 'season_name': '2015/2016'}    33
{'season_id': 24, 'season_name': '2012/2013'}    32
{'season_id': 25, 'season_name': '2013/2014'}    31
{'season_id': 41, 'season_name': '2008/2009'}    31
{'season_id': 40, 'season_name': '2007/2008'}    28
{'season_id': 39, 'season_name': '2006/2007'}    26
{'season_id': 38, 'season_name': '2005/2006'}    17
{'season_id': 37, 'season_name': '2004/2005'}     7
Name: season, dtype: int64

> we will retain only the season_name and discard the remaining data

In [14]:
#inspect home team and team
fcb[['home_team','away_team']].sample(3)

home_team  \
348  {'home_team_id': 213, 'home_team_name': 'Sevilla', 'home_team_gender': 'male', 'home_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 637, 'name': 'Jorge Luis Sampao...   
278  {'home_team_id': 223, 'home_team_name': 'Málaga', 'home_team_gender': 'male', 'home_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 188, 'name': 'Javier Gracia Carl...   
141  {'home_team_id': 395, 'home_team_name': 'Real Zaragoza', 'home_team_gender': 'male', 'home_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 4958, 'name': 'José Aurel...   

                                                                                                                                                                                                   away_team  
348  {'away_team_id': 217, 'away_team_name': 'Barcelona', 'away_team_gender': 'male', 'away_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 793, 'name': 'Luis Enrique Ma...  
278  {'away_team_id': 217, 'away_team_name': 'Barcelona', 'away_team_gender': 'male', 'away_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 793, 'name': 'Luis Enrique Ma...  
141  {'away_team_id': 217, 'away_team_name': 'Barcelona', 'away_team_gender': 'male', 'away_team_group': None, 'country': {'id': 214, 'name': 'Spain'}, 'managers': [{'id': 36, 'name': 'Josep Guardiola ...

> we will extract home_team_name and home_manager likewise away_team_name and away_team-manager

In [15]:
#inspect match_status
fcb['match_status'].value_counts()

available    520
Name: match_status, dtype: int64

> match status is unique, we will discard it

In [17]:
#inspect match_status_360
fcb['match_status_360'].value_counts()

scheduled     485
available      34
processing      1
Name: match_status_360, dtype: int64

In [18]:
#inspect meta data
fcb['metadata'].value_counts()

{'data_version': '1.1.0', 'shot_fidelity_version': '2', 'xy_fidelity_version': '2'}    465
{'data_version': '1.1.0', 'shot_fidelity_version': '2'}                                 26
{'data_version': '1.0.2', 'shot_fidelity_version': '2'}                                 16
{'data_version': '1.0.3', 'shot_fidelity_version': '2'}                                 13
Name: metadata, dtype: int64

> Not important for our analysis

In [19]:
#inspect competition stage
fcb['competition_stage'].value_counts()

{'id': 1, 'name': 'Regular Season'}    520
Name: competition_stage, dtype: int64

> has unique values, it shall be discarded

In [20]:
#inspect stadium
fcb['stadium'].value_counts()

{'id': 342, 'name': 'Camp Nou', 'country': {'id': 214, 'name': 'Spain'}}                                262
{'id': 361, 'name': 'Estadio de la Cerámica', 'country': {'id': 214, 'name': 'Spain'}}                   15
{'id': 353, 'name': 'Estadio Santiago Bernabéu', 'country': {'id': 214, 'name': 'Spain'}}                15
{'id': 344, 'name': 'Estadio de Mestalla', 'country': {'id': 214, 'name': 'Spain'}}                      15
{'id': 349, 'name': 'Estadio Ramón Sánchez Pizjuán', 'country': {'id': 214, 'name': 'Spain'}}            14
{'id': 5043, 'name': 'Reale Arena', 'country': {'id': 214, 'name': 'Spain'}}                             12
{'id': 354, 'name': 'Coliseum Alfonso Pérez', 'country': {'id': 214, 'name': 'Spain'}}                   11
{'id': 625, 'name': 'Estadio Vicente Calderón', 'country': {'id': 214, 'name': 'Spain'}}                 11
{'id': 351, 'name': 'RCDE Stadium', 'country': {'id': 214, 'name': 'Spain'}}                             10
{'id': 352, 'name': 'Estadio

> we will extract only the stadium name

In [21]:
#inspect referee
fcb['referee'].value_counts()

{'id': 221, 'name': 'Alberto Undiano Mallenco', 'country': {'id': 214, 'name': 'Spain'}}               33
{'id': 180, 'name': 'Antonio Miguel Mateu Lahoz', 'country': {'id': 214, 'name': 'Spain'}}             32
{'id': 222, 'name': 'David Fernández Borbalan', 'country': {'id': 214, 'name': 'Spain'}}               30
{'id': 1008, 'name': 'Carlos Clos Gómez', 'country': {'id': 214, 'name': 'Spain'}}                     20
{'id': 1003, 'name': 'Fernando Teixeira', 'country': {'id': 214, 'name': 'Spain'}}                     20
{'id': 211, 'name': 'José Luis González González', 'country': {'id': 214, 'name': 'Spain'}}            19
{'id': 183, 'name': 'Jesús Gil Manzano', 'country': {'id': 214, 'name': 'Spain'}}                      19
{'id': 1001, 'name': 'Miguel Pérez', 'country': {'id': 214, 'name': 'Spain'}}                          17
{'id': 2535, 'name': 'Carlos Velasco Carballo', 'country': {'id': 214, 'name': 'Spain'}}               17
{'id': 997, 'name': 'César Muñiz', 'country': 

> we will extract only the referee name

### Assessment
1. match_date, kick_off should be datetime
2. competition, competition_stafe, and match_status are unique (not important for our analysis)
3. season should have only the season_name (For example, 2004/2005)
4. home_team should consist only home_team name, also we can extract home_team manager from it.
5. away_team should consist only away_team name, also we can extract away_team manager from it.
6. last_updated, last_updated_360, and meta_data not needed
7. stadium should contain only the stadium_name
8. referee should contain only the referee_name.
9. create new column 'match_outcome' that have values of win, draw, lose for FCB.